src/interfaces/chain.cpp

```diff
@@ -185,8 +175,26 @@ class NotificationsHandlerImpl : public Handler, CValidationInterface
         m_notifications->updatedBlockTip();
     }
     void ChainStateFlushed(const CBlockLocator& locator) override { m_notifications->chainStateFlushed(locator); }
-    Chain& m_chain;
-    Chain::Notifications* m_notifications;
+    std::shared_ptr<Chain::Notifications> m_notifications;
+};
+
+class NotificationsHandlerImpl : public Handler
+{
+public:
+    explicit NotificationsHandlerImpl(std::shared_ptr<Chain::Notifications> notifications)
+        : m_proxy(std::make_shared<NotificationsProxy>(std::move(notifications)))
+    {
+        RegisterSharedValidationInterface(m_proxy);
+    }
+    ~NotificationsHandlerImpl() override { disconnect(); }
```

Comment:
nit: The destructor of the derived class should not be declared with `override`.

Reply:
Done.

Reply:
re: https://github.com/bitcoin/bitcoin/pull/18338#discussion_r399672974

> nit: The destructor of the derived class should not be declared with `override`.

This is a strange suggestion and I would suggest reverting 17316b3a6eb41204824111ea3cfeb2e1b7d80b1b to 41b0baf43c243b64b394e774e336475a489cca2b to undo it. The point of override keyword is to get the compiler to enforce assumptions the derived class is making about the base class to avoid terrible bugs. In this case, override keyword makes it an error for the base class not to have a virtual destructor. This is important because if base destructor was made not virtual there would be a difficult to trace memory leak here with NotificationsHandlerImpl object getting freed but destructor never running and m_proxy member never being freed.

Reply:
A desctructor is not a regular member function. So I was confused with the following opinions:
- https://github.com/isocpp/CppCoreGuidelines/issues/721
- https://github.com/isocpp/CppCoreGuidelines/pull/737

Reply:
@ryanofsky 
> In this case, override keyword makes it an error for the base class not to have a virtual destructor.

I've verified this statement with the following diff:
```diff
--- a/src/interfaces/handler.h
+++ b/src/interfaces/handler.h
@@ -22,7 +22,7 @@ namespace interfaces {
 class Handler
 {
 public:
-    virtual ~Handler() {}
+    ~Handler() {}
 
     //! Disconnect the handler.
     virtual void disconnect() = 0;
```

My compiler raises an error. You are correct.
TIL. Thank you.